### PR TITLE
Add subdomain support to dserve

### PR DESCRIPTION
**Pull request by griffbrad**
Current versions of dserve use a single domain for all commits and branches.  This can allow cross-contamination of client-side resources like cookies, localStorage and assets cached by service workers across different builds/images.

By instead using subdomains, we can leverage the browser's built-in mechanisms for isolating domains to ensure that our builds don't interact in unpredictable ways.

The new middleware in this commit maintains support for "hash" and "branch" query string parameters.  When present, those parameters will trigger a redirect to [commit-hash].[host-name].

The next middleware in the processing chain then grabs the commit hash from the subdomain and stores it in the session, where the rest of dserve expects to find it.

## Testing 

To test locally, you'll need hosts file entries for whichever commits you want to test:

`127.0.0.1 [commit-hash].calypso.localhost`

When accessing a branch or commit hash via query string parameter, you'll be redirected to the equivalent subdomain.  Navigating to `calypso.localhost:3000?hash=b84096f22f59c307620d4978bc48ea47da65d5a6`, for example, will redirect you to `b84096f22f59c307620d4978bc48ea47da65d5a6.calypso.localhost:3000`.

### CODE REVIEW

File: src/middlewares.ts [hunk 1]
```diff
@@ -15,37 +15,85 @@ import {
...
+    return protocol + '://hash-' + commitHash + '.' + stripCommitHashSubdomainFromHost(req.headers.host);
+}
 
-	if (!isHashSpecified) {
-		res.send('Please specify a branch to load');
-		return;
-	}
+function stripCommitHashSubdomainFromHost(host: string) {
+    let segments = host.split('.'),
+        commitHashIndex = null;
```
Comment (dmsnell) [inline src/middlewares.ts]: with `let` and `const` our code style implies that we can use one declaration per line, in contrast to `var` where we join them in a way that matches the hoisting going on with `var`

```js
const segments = host.split( '.' );
let commitHashIndex;
```

sorry for the drive-by nitpick. feel free to ignore.

File: src/middlewares.ts [hunk 2]
```diff
@@ -15,37 +15,85 @@ import {
...
+            break;
+        }
+    }
+
+    if (null === commitHashIndex) {
+        return host;
+    }
+
+    return segments.slice(commitHashIndex + 1).join('.');
+}
```
Comment (dmsnell) [inline src/middlewares.ts]: it would be helpful if we had a comment with some example input and output here as I'm finding the logic hard to parse in my head. what are we trying to accomplish here? are we trying to take `hash-12345.calypso.live/me` and produce `calypso.live/me` from it?

alternatively some unit tests could clarify the expected usage


we might be able to simplify the logic by removing some of the procedural steps and focus more on the semantic transformation?
```js
function withoutHashSubdomain( host ) {
	return host.replace( /(^|.*?\.)hash-[a-f0-9]+\./, '' );
}
```

I payed around with this in a jsbin - http://jsbin.com/cemogujami/1/edit?js,console - but I was unable to infer what the intended transformation was so the code could be wrong - you can verify by looking at the example input strings and see if the outputs produced by the `regexp` function are the correct transformed host strings.

File: src/middlewares.ts [hunk 3]
```diff
@@ -15,37 +15,85 @@ import {
...
+
+function getCommitHashFromSubdomain(req: any) {
+    const commitHash = _.find(
+        req.headers.host.split('.'),
+        function (hostSegment) {
+            return 0 === hostSegment.indexOf('hash-');
+        }
+    );
+
+    return commitHash.replace(/^hash-/, '');
```
Comment (dmsnell) [inline src/middlewares.ts]: this could give us a type error when trying to call `replace` on `undefined` if there's no match.

we could take the regexp approach as well…

```js
function commitHashFromHost( host: string ): ?string {
	const match = /(?:^|.*?\.)hash-([a-f0-9]+)\./.exec( host );

	if ( ! match ) {
		return null;
	}

	const [ /* full match */, hash ] = match;
	return hash;
}
```

Comment (dmsnell) [reply src/middlewares.ts]: is there a need to pass in `express.Request` here vs. a string, letting the caller unwrap it?

File: src/middlewares.ts [hunk 4]
```diff
@@ -15,37 +15,85 @@ import {
...
+        next();
+        return;
+    }
+
+    if (req.query.hash) {
 		commitHash = req.query.hash;
 	} else if (req.query.branch) {
 		commitHash = getCommitHashForBranch(req.query.branch);
-	}
+    }
```
Comment (dmsnell) [inline src/middlewares.ts]: fun fact! there's no need for mutable `let` bindings here…

```js
const commitHash = req.query.hash || getCommitHashForBranch( req.query.branch );
```

File: src/middlewares.ts [hunk 5]
```diff
@@ -15,37 +15,85 @@ import {
...
+        // @todo Render a nicer page here.
+		res.send('Please specify a branch to load');
 		return;
 	}
 
-	req.session.commitHash = commitHash;
-	touchCommit(commitHash);
+	req.session.commitHash = subdomainCommitHash;
+
+    touchCommit(subdomainCommitHash);
```
Comment (dmsnell) [inline src/middlewares.ts]: hm…might want to run `prettier` in your IDE - it looks like we have a tab vs. spaces issue here

File: src/middlewares.ts [hunk 6]
```diff
@@ -15,37 +15,85 @@ import {
...
-	const isHashSpecified = req.query && (req.query.hash || req.query.branch);
-	let commitHash;
+function assembleSubdomainUrlForHash(req: any, commitHash: string) {
+    const protocol = (req.secure ? 'https' : 'http');
 
-	if (isHashInSession && !isHashSpecified) {
-		next();
-		return;
-	}
+    return protocol + '://hash-' + commitHash + '.' + stripCommitHashSubdomainFromHost(req.headers.host);
```
Comment (blowery) [inline src/middlewares.ts]: why do we need the `hash-` prefix?

File: src/middlewares.ts [hunk 7]
```diff
@@ -15,37 +15,85 @@ import {
...
+    if (req.query.hash) {
 		commitHash = req.query.hash;
 	} else if (req.query.branch) {
 		commitHash = getCommitHashForBranch(req.query.branch);
-	}
+    }
 
-	if (commitHash instanceof Error) {
-		res.send(striptags('Calypso Server: ' + commitHash.message));
+    res.writeHead(
```
Comment (blowery) [inline src/middlewares.ts]: can probably just use response.redirect here https://expressjs.com/en/api.html#res.redirect

File: src/middlewares.ts [hunk 8]
```diff
@@ -15,37 +15,85 @@ import {
...
+        function (hostSegment) {
+            return 0 === hostSegment.indexOf('hash-');
+        }
+    );
+
+    return commitHash.replace(/^hash-/, '');
+}
 
-	if (req.query.hash) {
+export async function redirectHashFromQueryStringToSubdomain(req: any, res: any, next: any) {
```
Comment (blowery) [inline src/middlewares.ts]: why async? i don't see any `await` usage here.

File: src/middlewares.ts (hunk 6 above)
Comment (griffbrad) [reply src/middlewares.ts]: I explained the "hash-" prefix on this commit:
https://github.com/Automattic/dserve/pull/62/commits/8c164c42c8f2bd7ebb64edaf2d53b3e6c2062cf5
Basically, it seems like I need either a predictable/consistent format for the hash subdomain or a configuration file that specifies the base URL for the environment.

File: src/middlewares.ts (hunk 7 above)
Comment (griffbrad) [reply src/middlewares.ts]: Thanks for the pointer.  I think I was looking at core Node HTTP docs rather than Express.

File: src/middlewares.ts (hunk 8 above)
Comment (griffbrad) [reply src/middlewares.ts]: Removed it.  Looks like async was there for a while, so perhaps mapping branches to hashes was previously async?  At this point, that call is synchronous, though.

File: src/middlewares.ts (hunk 5 above)
Comment (griffbrad) [reply src/middlewares.ts]: I added a separate PR for a prettier config and pre-commit hook based upon the equivalents in Calypso:
https://github.com/Automattic/dserve/pull/63

File: src/middlewares.ts (hunk 4 above)
Comment (griffbrad) [reply src/middlewares.ts]: Noted!  Thanks!

File: src/middlewares.ts (hunk 3 above)
Comment (griffbrad) [reply src/middlewares.ts]: I'm now using a shared regex pattern between the two utility functions.

**Comment by sirreal** [comment]
Related work in D17735-code

**Comment by sirreal** [comment]
This PR adds both the hash-subdomain support and the redirection, is that correct?

In terms of testing, reviewing and landing –especially with D17735-code in mind– would it be simpler to split this into 2 changes?
1. Add _support_ for subdomain hashes (continue to support GET param hashes and branches, no redirection).
1. At this point D17735-code could be tested and landed with greater ease.
1. Add the redirections.

**Comment by griffbrad** [comment]
@sirreal You can actually test D17735-code prior to this PR being merged.  Ops folks have already configured the production dserve instance to support subdomains at the server level.  dserve just doesn't read/understand those subdomains yet.  

If you go to https://foo.calypso.live/?branch=master, for example, you'll remain on foo.calypso.live.  Without D17735-code, you'll get various auth issues.  With it, you'll be able to use Calypso normally.

Once this PR is merged, then "?branch=master" will be redirected to [commit-hash].calypso.live to achieve the improved build isolation.

**Comment by sirreal** [comment]
> You can actually test D17735-code prior to this PR being merged. Ops folks have already configured the production dserve instance to support subdomains at the server level.

Excellent, thanks for clearing that up 👍 

D17801-code does depend on this
